src/core/manager.c

```diff
@@ -3736,10 +3737,67 @@ int manager_deserialize(Manager *m, FILE *f, FDSet *fds) {
         return manager_deserialize_units(m, f, fds);
 }
 
+static int manager_pin_all_cgroup_bpf_programs(Manager *m, Set **pinned_progs) {
+        int r;
+        Unit *u;
+
+        assert(m);
+
+        HASHMAP_FOREACH(u, m->units) {
+                size_t i;
+                BPFProgram *p;
+                BPFProgram *builtin_progs[] = {u->bpf_device_control_installed,
+                                           u->ip_bpf_ingress,
+                                           u->ip_bpf_ingress_installed,
+                                           u->ip_bpf_egress,
+                                           u->ip_bpf_egress_installed};
+                Set *custom_progs[] = {u->ip_bpf_custom_ingress,
+                                       u->ip_bpf_custom_egress,
+                                       u->ip_bpf_custom_ingress_installed,
+                                       u->ip_bpf_custom_egress_installed};
+
+                for (i = 0; i < ELEMENTSOF(builtin_progs); i++) {
+                        p = builtin_progs[i];
+                        if (!p)
+                                continue;
+
+                        r = set_ensure_put(pinned_progs, NULL, p);
```

Comment:
We usually do not modify return values when the function fails. So, please do like this:
```
static int manager_pin_all_cgroup_bpf_programs(Manager *m, Set **pinned_progs) {
        _cleanup_set_free_ Set *set = NULL;
        Unit *u;
        int r;

        (snip)

        *pinned_progs = TAKE_PTR(set);
        return 0;
}
```
Moreover, on failure, each bpf program should be `unref()`ed. (So, `_cleanup_set_free_` above is not good...)

Or, `manager_reload()` says that `This is the point of no return, from here on there is no way back.` So, it may be OK to keep this function as is, but I'd like to add some comments about that in this function.

Reply:
> We usually do not modify return values when the function fails. So, please do like this.

Sure thing.

> Moreover, on failure, each bpf program should be unref()ed.

Hmm, could you please clarify? We already only `bpf_program_ref(p)` when the set put succeeded. All of the ones that were successfully put into the set will be unreferenced later in `unpin_all_cgroup_bpf_programs`, intermediate failure or not, no?

Reply:
>  All of the ones that were successfully put into the set will be unreferenced later in `unpin_all_cgroup_bpf_programs`, intermediate failure or not, no?

Yeah, it looks OK.